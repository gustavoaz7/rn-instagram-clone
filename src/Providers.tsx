import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider, ProviderProps } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { RootSiblingParent } from 'react-native-root-siblings'; // dep of rn-root-toast
import { theme as stylesTheme } from './styles/theme';
import { persistor, store as reduxStore } from './redux/store';
import { useThemeVariantSelector } from './redux/theme-variant';

type TProvidersProps = {
  children?: ReactNode;
  store?: ProviderProps['store'];
  theme?: typeof stylesTheme;
};

export function Providers({
  children = null,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <RootSiblingParent>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Theming theme={theme}>{children}</Theming>
        </PersistGate>
      </Provider>
    </RootSiblingParent>
  );
}

function Theming({
  theme,
  children,
}: Required<Omit<TProvidersProps, 'store'>>): JSX.Element {
  const themeVariant = useThemeVariantSelector();
  return <ThemeProvider theme={theme[themeVariant]}>{children}</ThemeProvider>;
}
