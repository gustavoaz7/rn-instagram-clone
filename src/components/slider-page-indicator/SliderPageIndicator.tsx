import React, { useEffect, useRef } from 'react';
import { ViewProps, Animated } from 'react-native';
import styled, { useTheme } from 'styled-components/native';

export type TSliderPageCounterProps = {
  total: number;
  current: number;
  style?: ViewProps['style'];
};

export const DELAY = 5000;

export const SliderPageIndicator = ({
  total,
  current,
}: TSliderPageCounterProps): JSX.Element => {
  const theme = useTheme();
  const opacity = useRef(new Animated.Value(1)).current;
  const initialPage = useRef(current).current;
  const hasChangedPageRef = useRef(false);

  useEffect(() => {
    if (current !== initialPage && !hasChangedPageRef.current) {
      hasChangedPageRef.current = true;
    }
  }, [current, initialPage]);

  useEffect(() => {
    const animations: Animated.CompositeAnimation[] = [
      Animated.timing(opacity, {
        toValue: 1,
        duration: 0,
        useNativeDriver: true,
      }),
    ];
    if (hasChangedPageRef.current) {
      animations.push(
        Animated.timing(opacity, {
          toValue: 0,
          duration: theme.animation.timingSlow,
          delay: DELAY,
          useNativeDriver: true,
        }),
      );
    }

    Animated.sequence(animations).start();
  }, [current, opacity, theme.animation.timingSlow]);

  return (
    <MediaCounterTag style={{ opacity }}>
      {current}/{total}
    </MediaCounterTag>
  );
};

const MediaCounterTag = styled(Animated.Text)`
  position: absolute;
  z-index: 1;
  ${({ theme: { spacing, color } }) => `
    top: ${spacing.m};
    right: ${spacing.m};
    padding: ${spacing.xs} ${spacing.s};
    border-radius: 99px;
    color: ${color.white};
    background-color: ${color.black}E6;
  `}
`;
