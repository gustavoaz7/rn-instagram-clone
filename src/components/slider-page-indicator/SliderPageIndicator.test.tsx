import React from 'react';
import { render } from '@testing-library/react-native';
import { SliderPageIndicator, DELAY } from './SliderPageIndicator';
import { Providers } from '../../Providers';
import {
  setupTimeTravel,
  destroyTimeTravel,
  timeTravel,
} from '../../test/time-travel';
import { defaultTheme } from '../../test/default-theme';

describe('components - SliderPageIndicator', () => {
  const options = { wrapper: Providers };
  const props = { total: 4, current: 3 };

  beforeEach(() => {
    setupTimeTravel();
  });

  afterEach(() => {
    destroyTimeTravel();
  });

  it('renders', () => {
    render(<SliderPageIndicator {...props} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<SliderPageIndicator {...props} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correct text', () => {
    const { getByText } = render(<SliderPageIndicator {...props} />, options);

    expect(getByText(`${props.current}/${props.total}`)).toBeTruthy();
  });

  it('renders correct text', () => {
    const { getByText } = render(<SliderPageIndicator {...props} />, options);

    expect(getByText(`${props.current}/${props.total}`)).toBeTruthy();
  });

  describe('animation', () => {
    it('starts visible', () => {
      const { getByText } = render(<SliderPageIndicator {...props} />, options);

      expect(getByText(`${props.current}/${props.total}`)).toHaveStyle({
        opacity: 1,
      });
    });

    describe('when stays on same page', () => {
      it('keeps visible', () => {
        const { getByText } = render(
          <SliderPageIndicator {...props} />,
          options,
        );

        timeTravel(DELAY * 2);

        expect(getByText(`${props.current}/${props.total}`)).toHaveStyle({
          opacity: 1,
        });
      });
    });

    describe('when changes page', () => {
      it(`hides after ${DELAY}ms`, () => {
        const { getByText, rerender } = render(
          <SliderPageIndicator total={4} current={2} />,
          options,
        );

        rerender(<SliderPageIndicator total={4} current={3} />);
        timeTravel(DELAY + defaultTheme.animation.timingSlow);

        expect(getByText('3/4')).toHaveStyle({ opacity: 0 });
      });

      describe('when go back to initial page', () => {
        it(`still hides after ${DELAY}ms`, () => {
          const { getByText, rerender } = render(
            <SliderPageIndicator total={4} current={2} />,
            options,
          );

          rerender(<SliderPageIndicator total={4} current={3} />);
          rerender(<SliderPageIndicator total={4} current={2} />);
          timeTravel(DELAY + defaultTheme.animation.timingSlow);

          expect(getByText('2/4')).toHaveStyle({ opacity: 0 });
        });
      });
    });
  });
});
