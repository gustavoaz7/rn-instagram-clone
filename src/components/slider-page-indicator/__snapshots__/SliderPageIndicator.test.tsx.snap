// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - SliderPageIndicator matches snapshot 1`] = `
<Text
  style={
    Object {
      "backgroundColor": "#020202e6",
      "borderRadius": 99,
      "color": "#FFFFFF",
      "opacity": 1,
      "paddingBottom": 5,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 5,
      "position": "absolute",
      "right": 15,
      "top": 15,
      "zIndex": 1,
    }
  }
>
  3
  /
  4
</Text>
`;
